### PR TITLE
misc: Change devcontainer for isca tutorial and bootcamp

### PR DESCRIPTION
**Pull request by powerjg**
Change-Id: I4e946bf32b5ad362ff34de5d354fe29cd26fc0cb

### CODE REVIEW

**Comment by powerjg** [comment]
Just for posterity to track the changes I made.

File: util/dockerfiles/devcontainer/Dockerfile [hunk 1]
```diff
@@ -39,15 +40,60 @@
...
 # This includes all dependencies for building and running gem5, the
 # dependencies for developing gem5, and a pre-built gem5 binary.
-FROM --platform=${BUILDPLATFORM} ghcr.io/gem5/ubuntu-24.04_all-dependencies
+FROM ghcr.io/gem5/ubuntu-22.04_all-dependencies:latest
 RUN apt -y update  && \
     apt -y install vim
-COPY --from=builder /gem5/build/ALL/gem5.opt /usr/local/bin/gem5
+    COPY --from=builder /gem5/build/ALL/gem5.opt /usr/local/bin/gem5-default
+    COPY --from=builder /gem5/build/ALL_CHI/gem5.opt /usr/local/bin/gem5
+    COPY --from=builder /gem5/build/VEGA_X86/gem5.opt /usr/local/bin/gem5-vega
```
Comment (BobbyRBruce) [inline util/dockerfiles/devcontainer/Dockerfile]: There is a weird leading whitespaces here.

```suggestion
COPY --from=builder /gem5/build/ALL/gem5.opt /usr/local/bin/gem5-default
COPY --from=builder /gem5/build/ALL_CHI/gem5.opt /usr/local/bin/gem5
COPY --from=builder /gem5/build/VEGA_X86/gem5.opt /usr/local/bin/gem5-vega
```

File: util/dockerfiles/devcontainer/Dockerfile [hunk 2]
```diff
@@ -39,15 +40,60 @@
...
+
+# Stage 3: Install rocm stuff
+RUN apt -y update && apt -y upgrade && \
+    apt -y install build-essential git m4 scons zlib1g zlib1g-dev \
+    libprotobuf-dev protobuf-compiler libprotoc-dev libgoogle-perftools-dev \
+    python3-dev python-is-python3 doxygen libboost-all-dev \
+    libhdf5-serial-dev python3-pydot libpng-dev libelf-dev pkg-config
+RUN apt -y install wget
```
Comment (BobbyRBruce) [inline util/dockerfiles/devcontainer/Dockerfile]:
```suggestion
    libhdf5-serial-dev python3-pydot libpng-dev libelf-dev pkg-config wget
```

File: util/dockerfiles/devcontainer/Dockerfile [hunk 3]
```diff
@@ -39,15 +40,60 @@
...
+
+RUN apt update
+
+# This package should be the minimum needed to build and is significantly
+# smaller than the full `rocm` package.
+RUN apt -y install rocm-dev
+
+# Stage 4: Install cross compilers
+RUN apt -y install gcc-aarch64-linux-gnu
+RUN apt -y install gcc-riscv64-linux-gnu
```
Comment (BobbyRBruce) [inline util/dockerfiles/devcontainer/Dockerfile]: 1. Best to run any `apt` command with the `-y` flag.
2. `apt` `update` and `install` commands should always be in a single `RUN`.
3. Where possible `install` commands should be merged to reduce the Docker image size.

```suggestion
RUN apt -y update && apt -y install \
    # This package should be the minimum needed to build and is significantly
    # smaller than the full `rocm` package.
    rocm-dev \
    # Installs the cross compilers.
    gcc-aarch64-linux-gnu \
    gcc-riscv64-linux-gnu
```

File: util/dockerfiles/devcontainer/Dockerfile [hunk 4]
```diff
@@ -39,15 +40,60 @@
...
+FROM ghcr.io/gem5/ubuntu-22.04_all-dependencies:latest as builder
+RUN git clone --branch release-staging-v24.0.0.0 https://github.com/gem5/gem5
 WORKDIR /gem5
 RUN scons build/ALL/gem5.opt -j`nproc`
+RUN scons defconfig build/ALL_CHI build_opts/ALL && \
+    scons setconfig build/ALL_CHI RUBY_PROTOCOL_CHI=y && \
+    scons build/ALL_CHI/gem5.opt -j`nproc`
+RUN scons build/VEGA_X86/gem5.opt -j`nproc`
 
 # Stage 2: The final stage where we create the devcontainer image.
```
Comment (BobbyRBruce) [inline util/dockerfiles/devcontainer/Dockerfile]:
```suggestion
# Stage 2: Create the devcontainer image.
```

File: util/dockerfiles/devcontainer/Dockerfile [hunk 5]
```diff
@@ -39,15 +40,60 @@
...
 # This includes all dependencies for building and running gem5, the
 # dependencies for developing gem5, and a pre-built gem5 binary.
```
Comment (BobbyRBruce) [inline util/dockerfiles/devcontainer/Dockerfile:53]:
```suggestion
# dependencies for developing gem5, and the pre-built gem5 binaries.
```

**Review by BobbyRBruce — CHANGES_REQUESTED** [review]

**Comment by BobbyRBruce** [comment]
Making some updates here. I'll let you know when ready for review.

File: util/dockerfiles/devcontainer/Dockerfile [hunk 6]
```diff
@@ -41,13 +42,33 @@
...
-    apt -y install vim
-COPY --from=builder /gem5/build/ALL/gem5.opt /usr/local/bin/gem5
+    apt -y install \
+    vim \
+    qemu-system \
+    qemu-kvm \
+    g++-aarch64-linux-gnu \
+    g++-riscv64-linux-gnu \
+    gcc-aarch64-linux-gnu \
+    gcc-riscv64-linux-gnu
```
Comment (powerjg) [inline util/dockerfiles/devcontainer/Dockerfile]: Can you add the following code? I'm not 100% sure what should go here vs in our startup script. This is necessary to get `menuconfig` to work.

```
apt-get update && apt-get install locales
locale-gen en_US.UTF-8
export LANG="en_US.UTF-8"
```

File: util/dockerfiles/devcontainer/Dockerfile [hunk 7]
```diff
@@ -41,13 +42,33 @@
...
+    scons build/ALL_CHI/gem5.opt -j`nproc`
+RUN scons build/VEGA_X86/gem5.opt -j`nproc`
 
-# Stage 2: The final stage where we create the devcontainer image.
+# Stage 2: Create the devcontainer image.
 # This includes all dependencies for building and running gem5, the
 # dependencies for developing gem5, and a pre-built gem5 binary.
 FROM --platform=${BUILDPLATFORM} ghcr.io/gem5/ubuntu-24.04_all-dependencies:latest
+COPY --from=builder /gem5/build/ALL/gem5.opt /usr/local/bin/gem5-mesi
+COPY --from=builder /gem5/build/ALL_CHI/gem5.opt /usr/local/bin/gem5
```
Comment (powerjg) [inline util/dockerfiles/devcontainer/Dockerfile]: Let's name this `gem5-chi` and then do a symlink from gem5 to gem5-chi

```suggestion
COPY --from=builder /gem5/build/ALL_CHI/gem5.opt /usr/local/bin/gem5-chi
RUN ln -s /usr/local/bin/gem5-chi /usr/local/bin/gem5
```

**Comment by BobbyRBruce** [comment]
Thanks Jason this PR and the dockerimage hosted on ghcr have been updated to incorporate this

**Review by BobbyRBruce — APPROVED** [review]

**Pull request merged**